src/typechecker/FStar.TypeChecker.Env.fs

```diff
@@ -175,10 +175,11 @@ and env = {
   nosynth        :bool;                         (* don't run synth tactics *)
   uvar_subtyping :bool;
   tc_term        :env -> term -> term*lcomp*guard_t; (* a callback to the type-checker; g |- e : M t wp *)
-  type_of        :env -> term -> term*typ*guard_t;   (* a callback to the type-checker; g |- e : Tot t *)
-  type_of_well_typed :env -> term -> typ;          (* a callback to the type-checker; falls back on type_of if the term is not well-typed *)
+  typeof_tot_or_gtot_term:env -> term -> bool -> term*typ*guard_t;   (* a callback to the type-checker; g |- e : Tot t *)
+  typeof_tot_or_gtot_term_fastpath  :env -> term -> option<typ>;       (* a callback to the type-checker, uses fast path *)
```

Comment:
From our discussions, I was expecting this to check and return whether the term is PURE or GHOST